### ISSUE TITLE
Different expressions with the same alias error on nested selects

### ISSUE THREAD

**Issue by JLeech** [yandex/ClickHouse#1777]
Hello, Clickhouse team!

I know that Clickhouse has function for Pearson Correlation, but had a bet to write it by my own.
Here it is: 
```
SELECT sum(mul)/sqrt(sum(sqr_dif_1) * sum(sqr_dif_2)) AS z_r
FROM(
SELECT 
        (SELECT avg(z_1) AS z_1_avg, 
                avg(z_2) AS z_2_avg
        FROM ( 
            SELECT perf_1.site, perf_1.z AS z_1
            FROM users.perf AS perf_1
            WHERE user_id = 000
        ) ALL INNER JOIN (
            SELECT perf_2.site, perf_2.z AS z_2
            FROM users.perf AS perf_2
            WHERE user_id = 999
        ) USING site) as avg_values,
       z_1 - avg_values.1 AS dif_1, 
       z_2 - avg_values.2 AS dif_2, 
       dif_1 * dif_2 AS mul, 
       dif_1*dif_1 AS sqr_dif_1, 
       dif_2*dif_2 AS sqr_dif_2
FROM (
            SELECT perf_1.site, perf_1.z AS z_1
            FROM users.perf AS perf_1
            WHERE user_id = 000
) ALL INNER JOIN (
            SELECT perf_2.site, perf_2.z AS z_2
            FROM users.perf AS perf_2
            WHERE user_id = 999
) USING site );
```
It calculates correlation of perfs.z for two users on same cite.

Sometimes it fails with such error:
```
Received exception from server:
Code: 179. DB::Exception: Received from localhost:9000, 127.0.0.1. DB::Exception: Different expressions with the same alias avg_values. 
0 rows in set. Elapsed: 11.947 sec.
``` 

ClickHouse client version 1.1.54310.
Connecting to localhost:9000.
Connected to ClickHouse server version 1.1.54310.
Table perf has 1'759'683'078 rows, with 234 sites
No other queries were evaluated during correlation count.

P.S.
this part never fails:
```
SELECT 
        (SELECT avg(z_1) AS z_1_avg, 
                avg(z_2) AS z_2_avg
        FROM ( 
            SELECT perf_1.site, perf_1.z AS z_1
            FROM users.perf AS perf_1
            WHERE user_id = 000
        ) ALL INNER JOIN (
            SELECT perf_2.site, perf_2.z AS z_2
            FROM users.perf AS perf_2
            WHERE user_id = 999
        ) USING site) as avg_values
```

**Comment by sleshJdev** [comment]
Any updates?

**Comment by alexey-milovidov** [comment]
Can you please provide an example table definition?

**Comment by alexey-milovidov** [comment]
```CREATE TABLE users.perf (site String, user_id UInt64, z Float64) ENGINE = Log```

**Comment by alexey-milovidov** [comment]
This is related to `enable_optimize_predicate_expression` setting:

```
milovidov-Latitude-E7470 :) SELECT * FROM (SELECT perf_1.z AS z_1 FROM users.perf AS perf_1)

SELECT *
FROM 
(
    SELECT perf_1.z AS z_1
    FROM users.perf AS perf_1 
) 

Received exception from server (version 18.11.0):
Code: 179. DB::Exception: Received from localhost:9000, 127.0.0.1. DB::Exception: Different expressions with the same alias z_1:
perf_1.z AS z_1
and
z AS z_1
. 

0 rows in set. Elapsed: 0.003 sec. 

milovidov-Latitude-E7470 :) SET enable_optimize_predicate_expression = 0

SET enable_optimize_predicate_expression = 0

Ok.

0 rows in set. Elapsed: 0.013 sec. 

milovidov-Latitude-E7470 :) SELECT * FROM (SELECT perf_1.z AS z_1 FROM users.perf AS perf_1)

SELECT *
FROM 
(
    SELECT perf_1.z AS z_1
    FROM users.perf AS perf_1 
) 

Ok.

0 rows in set. Elapsed: 0.013 sec.
```

@zhang2014 

It works perfectly if you disable it.

**Comment by alexey-milovidov** [comment]
#2015

**Comment by zhang2014** [comment]
It fixed in #3231 .
